todo-list/fn_and_obj.py

```diff
@@ -0,0 +1,129 @@
+        return ToDoList()
+    
+
+# Create class for list of ToDoItem, with required functionality
+class ToDoList:
+    
+    def __init__(self):
+        self.actual_list = []
+        
+    def add_item(self, item):
+         self.actual_list.append(item)
+        
+    def remove_item(self, item_r):
+        for i in range(len(self.actual_list)):
+            if self.actual_list[i].description == item_r:
+                del self.actual_list[i]
+                break
+        print('Item does not exist')
+                
+    # This function can inform whether an ToDoItem can be editted or if it doesn't exist
+    def item_exist(self, item_e):
+        for i in range(len(self.actual_list)):
+            if self.actual_list[i].description == item_e:
+                return True
+        return False 
+    
+    # Finding the index of a ToDoItem allows us to then call any item method.
+    # Otherwise each ToDoItem method would require a separate ToDoList method.
+    def find_item(self, item_f):
+        for i in range(len(self.actual_list)):
+            if self.actual_list[i].description == item_f:
+                return i
+        
+    # Each print option has a separate method (not condensed due to different printing)
+    def print_items(self):
+        for i in range(len(self.actual_list)):
+            if self.actual_list[i].complete:
+                completed_str = ' is completed'
+            else:
+                completed_str = ' is not completed'
+            print(self.actual_list[i].description + ' due on ' + 
+                      self.actual_list[i].date_str + completed_str)
+            
+    def print_incomp_items(self):
+        for i in range(len(self.actual_list)):
+            if not self.actual_list[i].complete:
+                print(self.actual_list[i].description + ' due on ' + 
+                      self.actual_list[i].date_str)
+    
+    def print_comp_items(self):
+        for i in range(len(self.actual_list)):
+            if self.actual_list[i].complete:
+                print(self.actual_list[i].description + ' due on ' + 
+                      self.actual_list[i].date_str)
+
+    def print_overdue_items(self):
+        for i in range(len(self.actual_list)):
+            if (not self.actual_list[i].complete) and datetime.now() > self.actual_list[i].date_obj:
+                print(self.actual_list[i].description + ' due on ' + 
+                      self.actual_list[i].date_str)
+
+    def print_search_items(self, search):
+        for i in range(len(self.actual_list)):
+            if search.lower() in self.actual_list[i].description.lower():
+                if self.actual_list[i].complete:
+                    completed_str = ' is completed'
+                else:
+                    completed_str = ' is not completed'
+                print(self.actual_list[i].description + ' due on ' + 
+                          self.actual_list[i].date_str + completed_str)
```

Comment:
Again, very generally speaking this way of constructing strings is fine, but can get very messy very quickly, for example what if you wanted to put an integer into your string, the following code would fail:

```
print("I am " + age + " years old")
```

There are a couple of better solutions to this, given you're using Python 3.6+ I cannot recommend [fstrings](https://realpython.com/python-f-strings/) highly enough.
For example in this case the code could be:

```
print(f'{self.actual_list[i].description} due on {self.actual_list[i].date_str} {completed_str}')
```

Reply:
Additionally, you have repeated the same code to generate a 'stringified' version of the TodoItem in the ToDoList quite a few times.
If you wanted to change the way that it was rendered you'd have to do this in all places.

Instead, what you might want to consider is using creating a `__str__()` method on the ToDoItem.
This is a bit of a 'magic' method, because what it will do is return a string representation of your object whenever it is converted to a string or printed.

For example:

```
class ToDoItem:
    def __str__(self):
        return f'{self.actual_list[i].description} due on {self.actual_list[i].date_str} {completed_str}'
```

Then you can just do this to print the item and you'll get the string representation you specified:

```
print(item)
```

You can read more about that in https://www.journaldev.com/22460/python-str-repr-functions.